api/src/main/java/io/druid/timeline/DataSegment.java

```diff
@@ -116,17 +116,18 @@ public boolean apply(String input)
     };
 
     // dataSource, dimensions & metrics are stored as canonical string values to decrease memory required for storing large numbers of segments.
-    this.dataSource = interner.intern(dataSource);
+    this.dataSource = STRING_INTERNER.intern(dataSource);
     this.interval = interval;
     this.loadSpec = loadSpec;
     this.version = version;
+
     this.dimensions = dimensions == null
                       ? ImmutableList.<String>of()
-                      : ImmutableList.copyOf(Iterables.transform(Iterables.filter(dimensions, nonEmpty), internFun));
+                      : internFn(Lists.newArrayList(Iterables.filter(dimensions, nonEmpty)));
     this.metrics = metrics == null
                    ? ImmutableList.<String>of()
-                   : ImmutableList.copyOf(Iterables.transform(Iterables.filter(metrics, nonEmpty), internFun));
-    this.shardSpec = (shardSpec == null) ? new NoneShardSpec() : shardSpec;
+                   : internFn(Lists.newArrayList(Iterables.filter(metrics, nonEmpty)));
```

Comment:
these could be `Ordering.natural().immutableSortedCopy(...)`, it incurs fewer copies according to the Ordering docs.